### ISSUE TITLE
Error in cuda flash attention when building cargo project

### ISSUE THREAD

**Issue by joshpopelka20** [EricLBuehler/mistral.rs#678]
## Minimum reproducible example
CUDA_NVCC_FLAGS="-fPIC" cargo run --release --features "cuda flash-attn cudnn" -- --token-source "literal:hf_..." -n "0:8;1:8;2:8;3:8" -i plain -m meta-llama/Meta-Llama-3.1-8B-Instruct -a llama

## Error
```
 #$ gcc -std=c++17 -D__CUDA_ARCH__=860 -D__CUDA_ARCH_LIST__=860 -c -x c++  -DCUDA_DOUBLE_MATH_FUNCTIONS -fPIC -O3 -I"cutlass/include" "-I/usr/local/cuda-12.1/bin/../targets/x86_64-linux/include"   -m64 "/tmp/tmpxft_00008798_00000000-6_flash_fwd_hdim96_bf16_causal_sm80.cudafe1.cpp" -o "/home/ec2-user/mistral.rs/target/release/build/candle-flash-attn-dab624edced762ec/out/flash_fwd_hdim96_bf16_causal_sm80-800cc4806a424737.o" 
  kernels/flash_fwd_launch_template.h: In lambda function:
  kernels/flash_fwd_launch_template.h:174:149: error: declaration of ‘constexpr const bool Is_causal’ shadows template parameter
       BOOL_SWITCH(params.p_dropout < 1.f, Is_dropout, [&] {
                                                                                                                                                       ^        
  kernels/flash_fwd_launch_template.h:171:20: note: template parameter ‘Is_causal’ declared here
   template<typename T, bool Is_causal>
                      ^~~~
  kernels/flash_fwd_launch_template.h:174:500: error: declaration of ‘constexpr const bool Is_causal’ shadows template parameter
       BOOL_SWITCH(params.p_dropout < 1.f, Is_dropout, [&] {
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      ^        
  kernels/flash_fwd_launch_template.h:171:20: note: template parameter ‘Is_causal’ declared here
   template<typename T, bool Is_causal>
                      ^~~~
  kernels/flash_fwd_launch_template.h: In lambda function:
  kernels/flash_fwd_launch_template.h:174:952: error: declaration of ‘constexpr const bool Is_causal’ shadows template parameter
       BOOL_SWITCH(params.p_dropout < 1.f, Is_dropout, [&] {
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          ^        
  kernels/flash_fwd_launch_template.h:171:20: note: template parameter ‘Is_causal’ declared here
   template<typename T, bool Is_causal>
                      ^~~~
  kernels/flash_fwd_launch_template.h:174:1303: error: declaration of ‘constexpr const bool Is_causal’ shadows template parameter
       BOOL_SWITCH(params.p_dropout < 1.f, Is_dropout, [&] {
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ^        
  kernels/flash_fwd_launch_template.h:171:20: note: template parameter ‘Is_causal’ declared here
   template<typename T, bool Is_causal>
                      ^~~~
  # --error 0x1 --
  thread '<unnamed>' panicked at /home/ec2-user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bindgen_cuda-0.1.5/src/lib.rs:262:21:
  nvcc error while executing compiling: "nvcc" "--gpu-architecture=sm_86" "-c" "-o" "/home/ec2-user/mistral.rs/target/release/build/candle-flash-attn-dab624edced762ec/out/flash_fwd_hdim96_bf16_causal_sm80-800cc4806a424737.o" "--default-stream" "per-thread" "-std=c++17" "-O3" "-U__CUDA_NO_HALF_OPERATORS__" "-U__CUDA_NO_HALF_CONVERSIONS__" "-U__CUDA_NO_HALF2_OPERATORS__" "-U__CUDA_NO_BFLOAT16_CONVERSIONS__" "-Icutlass/include" "--expt-relaxed-constexpr" "--expt-extended-lambda" "--use_fast_math" "--verbose" "--compiler-options" "-fPIC" "kernels/flash_fwd_hdim96_bf16_causal_sm80.cu"
```

## Other information
Please specify:
- Operating system:
```
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2"
HOME_URL="https://amazonlinux.com/"
SUPPORT_END="2025-06-30"
```
```
- GPU or accelerator information
    - If CUDA, please run `nvcc --version`, `nvidia-smi`
```
sh-4.2$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Mon_Apr__3_17:16:06_PDT_2023
Cuda compilation tools, release 12.1, V12.1.105
Build cuda_12.1.r12.1/compiler.32688072_0


![image](https://github.com/user-attachments/assets/d9cedbdc-9768-4072-bb70-b4c6bc82173e)


## Latest commit or version
[8cab33b](https://github.com/EricLBuehler/mistral.rs/commit/8cab33bb91096ab38e19565ceee75cdd52f0b02a)
 ·

**Comment by joshpopelka20** [comment]
I see in the code when the flag was added. Not sure why it's an issue now.
![image](https://github.com/user-attachments/assets/96910d8d-f590-4df6-8e71-ea67b237da03)


![image](https://github.com/user-attachments/assets/ab87f852-bd56-4c93-917c-7b2c8ec41d57)

**Comment by joshpopelka20** [comment]
I've added a PR that seems to fix this issue [PR #16](https://github.com/EricLBuehler/candle/pull/16). It looks like it was a merge issue from the code I previously fixed.

Even after that fix, there is another issue. When loading the model, I get this error:
```
2024-08-13T15:14:23.504107Z  INFO mistralrs_core::device_map: Layer 28: cuda[3]
2024-08-13T15:14:23.504109Z  INFO mistralrs_core::device_map: Layer 29: cuda[3]
2024-08-13T15:14:23.504112Z  INFO mistralrs_core::device_map: Layer 30: cuda[3]
2024-08-13T15:14:23.504116Z  INFO mistralrs_core::device_map: Layer 31: cuda[3]
Error: cannot find tensor lm_head.bias
```

**Comment by EricLBuehler** [comment]
Sounds good! I found the issue re lm_head.bias and will push a fix soon.

**Comment by EricLBuehler** [comment]
@joshpopelka20 this should be fixed in #681!

**Comment by joshpopelka20** [comment]
Thanks! It works now. I'll close the issue.
